baseplate/lock.py

```diff
@@ -0,0 +1,162 @@
+"""Simple locks in memcached or redis."""
+from __future__ import absolute_import
+
+import os
+import random
+import socket
+import time
+
+from pymemcache.client.base import PooledClient
+from redis import StrictRedis
+
+
+class TimeoutExpired(Exception):
+    pass
+
+
+MIN_RETRY_WAIT = 0.1
+MAX_RETRY_WAIT = 1.0
+
+
+class WrappedMemcache(object):
+    def __init__(self, cache):
+        self.cache = cache
+
+    def add_if_unset(self, key, value, ttl):
+        did_add = self.cache.add(key, value, expire=ttl, noreply=False)
+        return did_add
+
+    def get(self, key):
+        return self.cache.get(key)
+
+    def delete(self, key):
+        self.cache.delete(key, noreply=False)
+
+
+class WrappedRedis(object):
+    def __init__(self, cache):
+        self.cache = cache
+
+    def add_if_unset(self, key, value, ttl):
+        did_add = self.cache.set(key, value, nx=True, ex=ttl)
+
+        # StrictRedis returns True for 'set', and None for 'not set'
+        return bool(did_add)
+
+    def get(self, key):
+        return self.cache.get(key)
+
+    def delete(self, key):
+        self.cache.delete(key)
+
+
+class CacheLock(object):
+    """A lock implemented on memcached or redis.
+
+    Use the cache operation to add a key only if it doesn't exist. If
+    the key is not present in the cache its value is set and the server
+    responds that it was set. If the key is already present in the cache its
+    value is not modified and the server responds that it was not set.
+
+    We use this operation to ensure that only one client can obtain the lock
+    at a time. The first client gets the "set" response and interprets that to
+    mean that the lock was acquired. Any subsequent clients will get the
+    "not set" response and interpret that to mean that some other process has
+    the lock. They can wait and try again, or raise an exception.
+
+    """
+
+    def __init__(self, wrapped_cache, key, hold_limit, wait_limit):
+        self.wrapped_cache = wrapped_cache
+        self.key = key
+        self.hold_limit = hold_limit
+        self.wait_limit = wait_limit
+        self.have_lock = False
+
+    def acquire(self):
```

Comment:
Rather than this object having multiple states to juggle, could `acquire()` return an object that has `release()` on it? That way you can't even try to do it the wrong way as the methods won't exist.